[Refactor] 닉네임 관련 보안 이슈 설정

BAEK0111 opened the pull request:
## #️⃣연관된 이슈
> #165 

## 📝작업 내용
> 카카오 로그인 회원가입 로직 api 2개로 분리

## 🔎코드 설명
> 사용자 임시저장 : open id만 입력받고, 사용자의 상태를 PENDING 상태로 저장.
> 닉네임과 open id 입력받은 후 provider id를 통해 사용자를 찾고, 닉네임 및 사용자의 상태 PENDING -> ACTIVE로 업데이트

## 💬고민사항 및 리뷰 요구사항 (Optional)
> open Id를 입력받고, 사용자 PENDING상태로 저장할 때 access token, refresh token이 필요한지 모르겠습니다..! 제 생각에는 삭제해도 무방할 것 같은데 혹시 몰라서 남겨둠요..
또 이 방법으로 하면 gender, username, email, password 등등 null로 들어가는 값들이 많은데 이 부분 처리를 어떻게할지 같이 생각해봐주시면 감사하겠습니다!!
--------------------------------------------------
아래 사진설명
처음 2개 :  사용자 임시 저장 성공 & 상태 pending으로 들어감
나중 2개 : open id와 닉네임 입력받은 후 사용자 닉네임 & 상태 업데이트

## 비고 (Optional)
> 
<img width="1710" alt="스크린샷 2025-05-05 오후 10 16 02" src="https://github.com/user-attachments/assets/d5804f22-b96f-4b7f-a7c3-40a1ff0ff830" />
<img width="1710" alt="스크린샷 2025-05-05 오후 10 16 22" src="https://github.com/user-attachments/assets/d1ba9388-e6db-4ee5-9ea7-f8055043c198" />
<img width="1710" alt="스크린샷 2025-05-05 오후 10 17 36" src="https://github.com/user-attachments/assets/a2b3f204-2913-4d53-8d76-1322ce430f0b" />
<img width="1710" alt="스크린샷 2025-05-05 오후 10 17 46" src="https://github.com/user-attachments/assets/c991bdb5-fb0e-4c00-93c5-ffe2263b2639" />

## 최종 사진
>
1. open ID 검증 완료
<img width="1710" alt="openId 검증" src="https://github.com/user-attachments/assets/913f6de3-f1a4-4a64-a511-cb54b31ff44f" />
2. PENDING 상태 확인 -> true
<img width="1710" alt="PENDING상태 검증" src="https://github.com/user-attachments/assets/c76d2f80-c372-42c4-8e5a-280567b5d53a" />
3. 닉네임 입력 후 최종 회원가입
<img width="1710" alt="닉네임 입력 후 최종 회원가입" src="https://github.com/user-attachments/assets/ff394145-a534-4511-a877-a7b49e268d78" />
4. 최종 회원가입 후 PENDING -> false
<img width="1710" alt="최종 회원가입 후 PENDING 검증" src="https://github.com/user-attachments/assets/448c12b8-c6c3-41b6-9f26-467c923f10c3" />

BAEK0111 committed:
refactor: #165 H2 UserStatus Pending 추가, nickname null가능하게 변경

BAEK0111 committed:
refactor: #165 카카오 회원가입 API 분리
1. users/signup/kakao
open id를 입력받고, status를 PENDING 상태로 저장
2. users/signup/kakao/nickname
open id와 닉네임을 입력받아 해당 사용자를 찾은 후 닉네임과 상태 업데이트

yoonsu417 commented:
openId로 유저 구별만 가능하다면, 두개로 구별해서 처리하는 게 좋은 것 같습니다 :)
추가로 accesstoken, refreshtoken은 pending 상태에서는 불필요하다고 생각합니다. 아직 가입이 되지 않은 사용자니까 권한 부여의 필요가 없으니까요 !! ㅎㅎ

yoonsu417 commented:
프론트에서 openId..? 토큰..? 생성할 때, 카카오톡에서 정보 몇개 선택해서 넣을 수 있는걸로 알아요..!! 프론트랑 상의해서 정보 뽑아올거면 가져와서 db에 넣고 아니면 db에서 항목 자체를 삭제하는 식으로 가는게 좋을 것 같습니다 :)

Jinho622 commented:
고민사항 및 리뷰 요구사항 (Optional)
1. open Id를 입력받고, 사용자 PENDING상태로 저장할 때 access token, refresh token이 필요한지 모르겠습니다..! 제 생각에는 삭제해도 무방할 것 같은데 혹시 몰라서 남겨둠요..
-> 윤수님말대로 PENDING상태일때는 access token과 refresh token이 필요없습니다! 2개의 token은 인가처리를 할 때 필요한 것이기 때문에 카카오 로그인 또는 카카오 회원가입(카카오 회원가입은 로그인을 다시 하지 않고 바로 로그인 처리 되기 때문에 회원가입 후에도 토큰 발급) 이후에 발급해주면 됩니다!

2. 또 이 방법으로 하면 gender, username, email, password 등등 null로 들어가는 값들이 많은데 이 부분 처리를 어떻게할지 같이 생각해봐주시면 감사하겠습니다!!
-> username과 password는 일반 로그인에서만 사용하기 때문에 null이어야하고 email 또한 ID와 비밀번호찾기에만 사용되기 때문에 null이어도 상관없을 것 같습니다. gender도 당장 서비스에서는 활용되지 않기 때문에 상관없지만 필요하다면 카카오 서버에서 가져올 수 있습니다! 그리고 만약 null값이 많은 것이 리소스 낭비가 된다거나 다른 문제가 있다면 다른 방법을 생각해봐야 할텐데 이부분은 한번 알아보겠습니다.

Jinho622 commented on src/main/java/TtattaBackend/ttatta/service/UserService/UserCommandServiceImpl.java:
```diff
@@ -98,25 +98,60 @@ public Users signUp(UserRequestDTO.SignUpRequestDTO request) {
+    // 여기서 일상 카테고리 만들어도 될듯!
+    @Override
+    @Transactional
+    public UserResponseDTO.KaKaoFinalSignUpResultDTO kakaoSignUp(String openId, UserRequestDTO.SignUpKakaoRequestDTO request) {
```
User의 Status가 PENDING인지 한번 확인하는 작업이 들어가면 좋을 것 같아요!

Jinho622 commented on src/main/java/TtattaBackend/ttatta/web/controller/UserController.java:
```diff
@@ -97,36 +96,31 @@ public ApiResponse<?> logout(
-    // 미구현
-    @Operation(summary = "카카오 회원가입", description =
-            "# 카카오 회원가입 API 입니다. header에 'OpenId: {ID token}'형식으로 ID token을 입력하고 request body에 닉네임을 입력해주세요."
+    @Operation(summary = "카카오 openId 검증 API", description =
```
이 api는 없어도 될 거 같다 생각해요. 기존에 있던 api인 카카오 로그인 시 회원가입인지 로그인인지 확인하는 api(/users/verificate/kakao)에서 한번에 처리하면 될 거 같은데 어떻게 생각하시나요?? 앞서 말씀드린 기존 api에서 db에 해당 카카오 회원이 없다면(회원가입 상황이라면) status가 pending인 유저를 만들어두면 될 거 같아요

Jinho622 replied:
그리고 기존에 있던 api인 카카오 로그인 시 회원가입인지 로그인인지 확인하는 api(/users/verificate/kakao)에서 db에 있는 유저라도 status가 PENDING인지 확인하고 PENDING이라면 응답으로 반환해주면서 PENDING상태임을 알려줘야 될 것 같아요

Jinho622 commented:
+ 정정할 부분 있습니다!

PENDING상태로 유저가 생성됐을 때도 access token과 refresh token이 필요할 것 같습니다. 그리고 닉네임을 입력받았을때 요청하는 api에 openId대신 토큰을 전달하는게 나을 거 같아요(닉네임 설정 중간에 앱을 나간상황에서 다시 앱을 실행했을 때 카카오 로그인을 다시 하지 않아도 유저의 상태가 PENDING임을 확인하기 위해서)

BAEK0111 committed:
refactor: #165 카카오 회원가입 API 분리
1. ErrorStatus open Id 에서 sub 추출 시 없을 경우 에러 추가
2. openIdKakao api 와 validate token api 합침
- sub 추출 실패 시 에러처리
- sub 추출 성공 시
    - sub에 해당하는 user 존재 시 : isRegisterd = ture 제공, access, refresh token 제공하며 로그인 처리
    - sub에 해당하는 user 없을 때 : isRegistered = false 제공, access, refresh token 제공하고 pending 상태로 임시 회원가입 처리

BAEK0111 committed:
refactor: #165 카카오 회원가입 API 분리
1. SecurityUtil -> anonymous user 에러 처리
2. validateToken api 주석처리
3. 닉네임 입력받아 최종 회원가입하는 api에서 accessToken으로 사용자 구분하기 (아직 에러남)

Jinho622 committed:
fix: #165 'java.lang.RuntimeException: Security Context 에 인증 정보가 없습니다.' exception 해결

exception 발생 원인: JwtAuthenticationFilter에서 jwt토큰 검증히 필요하지 않은 api를 whitelist에 넣어 검증을 건너뛰었는데 whitelist api중 '/users/signup/**'가 포함돼있어
기존 카카오 회원가입 api(/users/signup/kakao/nickname)이 spring security의 jwt 토큰 검증을 건너뛰어 해당 에러가 발생했던 것임
해결 방법: 카카오 회원가입 api의 uri를 /users/kakao/signup/nickname으로 변경

exception 해결 외 수정사항: controller에서 accesstoken받을 필요 없음. controller에 데이터가 도달하기 전 spring security에서 header의 토큰을 가로채 검증을 수행함.
그리고 카카오 회원가입 api에서 access token과 refresh token을 새로 발급할 필요는 없어보이지만 새로 발급하여 응답에 넣어준다면 request로 받은 access token을 그대로 응답으로
전달하는 것이 아닌 refresh token처럼 새로 발급해야 될 것 같음.(service에서 spring security에서 가로챈 access token을 사용할 수 있다면 기존 코드대로 그대로 반환해주어도 될 것 같음)

BAEK0111 committed:
refactor: #165 추가 수정
- Pending 상태로 7일 지나면 자정에 삭제 구현

BAEK0111 committed:
refactor: #165 추가 수정
- 닉네임 입력 후 로그인처리 api 수정
  (access token, nickname) -> (nickname)
- JwtUtils.java 에 사용하지 않은 함수 삭제
  (제가 쓰려고 만든거였어요. 원래 있던거 안건들임)

BAEK0111 committed:
refactor: #165 추가 수정
- PENDING 상태인지 확인하는 API 구현
